As a user, I would appreciate some feedback when the declared time_zone_offset for a data source conflicts with the data inside that source

Given the declaration of a data source that contains a time_zone_offset
When that source refers to a time-series data format in which the time zone information is always explicit (integral)
Then the software should produce an error when the declaration is inconsistent with the data

See #491 for context.

@gabelewis2004

Now get this from the PI-XML reader:

```
Caused by: wres.reading.ReadException: The declared 'time_zone_offset' for the data source at 'file:///foo/data/DOSC1_QIN.xml' was -06:00, which does not match the 'timeZone' of 'Z' for a time-series within the source. Please resolve this conflict and try again.
```

And this from the CSV reader:

```
Caused by: wres.reading.ReadException: The declared 'time_zone_offset' for the data source at 'file:///foo/smalldata/pizza_near_volcano_observed_in_celsius.csv' was '-06:00', which is inconsistent with the CSV format requirement that all times are supplied in UTC. Please resolve this conflict and try again.
```

--- a/wres-reading/src/wres/reading/csv/CsvReader.java
+++ b/wres-reading/src/wres/reading/csv/CsvReader.java
@@ -4,6 +4,7 @@ import java.io.IOException;
 import java.io.InputStream;
 import java.time.Duration;
 import java.time.Instant;
+import java.time.ZoneOffset;
 import java.time.format.DateTimeParseException;
 import java.time.temporal.ChronoUnit;
 import java.util.EnumMap;
@@ -80,8 +81,8 @@ public class CsvReader implements TimeSeriesReader
     {
         Objects.requireNonNull( dataSource );
 
-        // Validate the disposition of the data source
-        ReaderUtilities.validateDataDisposition( dataSource, DataDisposition.CSV_WRES );
+        // Validate the data source
+        this.validateDataSource( dataSource );
 
         try
         {
@@ -101,8 +102,8 @@ public class CsvReader implements TimeSeriesReader
         Objects.requireNonNull( dataSource );
         Objects.requireNonNull( inputStream );
 
-        // Validate the disposition of the data source
-        ReaderUtilities.validateDataDisposition( dataSource, DataDisposition.CSV_WRES );
+        // Validate the data source
+        this.validateDataSource( dataSource );
 
         try
         {
@@ -534,6 +535,34 @@ public class CsvReader implements TimeSeriesReader
         }
 
         return ensembleName;
+    }
+
+    /**
+     * Validates the data source.
+     *
+     * @param dataSource the data source
+     */
+
+    private void validateDataSource( DataSource dataSource )
+    {
+        // Validate the disposition of the data source
+        ReaderUtilities.validateDataDisposition( dataSource, DataDisposition.CSV_WRES );
+
+        if ( Objects.nonNull( dataSource.getSource() )
+             && Objects.nonNull( dataSource.getSource()
+                                           .timeZoneOffset() )
+             && !Objects.equals( dataSource.getSource()
+                                           .timeZoneOffset(),
+                                 ZoneOffset.UTC ) )
+        {
+            throw new ReadException( "The declared 'time_zone_offset' for the data source at '"
+                                     + dataSource.getUri()
+                                     + "' was '"
+                                     + dataSource.getSource()
+                                                 .timeZoneOffset()
+                                     + "', which is inconsistent with the CSV format requirement that all times are "
+                                     + "supplied in UTC. Please resolve this conflict and try again." );
+        }
     }
 
     /**

--- a/wres-reading/src/wres/reading/fews/PublishedInterfaceXmlReader.java
+++ b/wres-reading/src/wres/reading/fews/PublishedInterfaceXmlReader.java
@@ -480,17 +480,19 @@ public final class PublishedInterfaceXmlReader implements TimeSeriesReader
         {
             ZoneOffset configuredOffset = dataSource.getSource()
                                                     .timeZoneOffset();
+
+            // Render this exceptional: GitHub 494
             if ( Objects.nonNull( configuredOffset )
                  && !configuredOffset.equals( ingestedOffset ) )
             {
-                LOGGER.warn( "The declared 'time_zone_offset' for the data source at '{}' was {}, which does not match "
-                             + "the 'timeZone' of {} for a time-series within the source. It is best not to declare "
-                             + "the 'time_zone_offset' for a PI-XML source in the project declaration when the "
-                             + "'timeZone' is available within the data source itself because the declaration will be "
-                             + "ignored.",
-                             dataSource.getUri(),
-                             configuredOffset,
-                             ingestedOffset );
+                throw new ReadException( "The declared 'time_zone_offset' for the data source at '"
+                                         + dataSource.getUri()
+                                         + "' was '"
+                                         + configuredOffset
+                                         + ", which does not match the 'timeZone' of '"
+                                         + ingestedOffset
+                                         + "' for a time-series within the source. Please resolve this conflict and try "
+                                         + "again." );
             }
         }
 

--- a/wres-reading/test/wres/reading/csv/CsvReaderTest.java
+++ b/wres-reading/test/wres/reading/csv/CsvReaderTest.java
@@ -1,6 +1,8 @@
 package wres.reading.csv;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
 
 import java.io.BufferedInputStream;
 import java.io.BufferedWriter;
@@ -10,6 +12,7 @@ import java.nio.file.FileSystem;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.time.Instant;
+import java.time.ZoneOffset;
 import java.util.Collections;
 import java.util.List;
 import java.util.Map;
@@ -22,6 +25,7 @@ import org.mockito.Mockito;
 import com.google.common.jimfs.Configuration;
 import com.google.common.jimfs.Jimfs;
 
+import wres.config.yaml.components.SourceBuilder;
 import wres.config.yaml.components.Variable;
 import wres.datamodel.MissingValues;
 import wres.datamodel.space.Feature;
@@ -29,6 +33,7 @@ import wres.datamodel.time.Event;
 import wres.datamodel.time.TimeSeries;
 import wres.datamodel.time.TimeSeriesMetadata;
 import wres.reading.DataSource;
+import wres.reading.ReadException;
 import wres.reading.TimeSeriesTuple;
 import wres.reading.DataSource.DataDisposition;
 import wres.statistics.MessageUtilities;
@@ -451,6 +456,54 @@ class CsvReaderTest
 
                 assertEquals( expected, actual );
             }
+
+            // Clean up
+            if ( Files.exists( csvPath ) )
+            {
+                Files.delete( csvPath );
+            }
+        }
+    }
+
+    @Test
+    void testReadObservationsWithInconsistentDeclaredTimeZoneProducesExpectedException() throws IOException
+    {
+        try ( FileSystem fileSystem = Jimfs.newFileSystem( Configuration.unix() ) )
+        {
+            // Write a new csv file to an in-memory file system
+            Path directory = fileSystem.getPath( TEST );
+            Files.createDirectory( directory );
+            Path pathToStore = fileSystem.getPath( TEST, TEST_CSV );
+            Path csvPath = Files.createFile( pathToStore );
+
+            try ( BufferedWriter writer = Files.newBufferedWriter( csvPath ) )
+            {
+                writer.append( "value_date,variable_name,location,measurement_unit,value\n" )
+                      .append( "1985-06-01T13:00:00Z,QINE,DRRC2,CFS,1\n" )
+                      .append( "1985-06-01T14:00:00Z,QINE,DRRC2,CFS,\n" )
+                      .append( "1985-06-01T15:00:00Z,QINE,DRRC2,CFS,3" );
+            }
+
+            DataSource dataSource = Mockito.mock( DataSource.class );
+            Mockito.when( dataSource.getUri() )
+                   .thenReturn( csvPath.toUri() );
+            Mockito.when( dataSource.getVariable() )
+                   .thenReturn( new Variable( QINE, null, Set.of() ) );
+            Mockito.when( dataSource.getDisposition() )
+                   .thenReturn( DataDisposition.CSV_WRES );
+            Mockito.when( dataSource.getSource() )
+                   .thenReturn( SourceBuilder.builder()
+                                             .timeZoneOffset( ZoneOffset.of( "-06:00" ) )
+                                             .build() );
+
+            CsvReader reader = CsvReader.of();
+
+            // Validation happens on stream construction, even if reading happens later
+            ReadException actual = assertThrows( ReadException.class,
+                                                 () -> reader.read( dataSource ) );
+            assertTrue( actual.getMessage()
+                              .contains( "which is inconsistent with the CSV format requirement that all times are "
+                                         + "supplied in UTC" ) );
 
             // Clean up
             if ( Files.exists( csvPath ) )

--- a/wres-reading/test/wres/reading/fews/PublishedInterfaceXmlReaderTest.java
+++ b/wres-reading/test/wres/reading/fews/PublishedInterfaceXmlReaderTest.java
@@ -1,6 +1,8 @@
 package wres.reading.fews;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
 
 import java.io.BufferedInputStream;
 import java.io.BufferedWriter;
@@ -33,6 +35,7 @@ import wres.datamodel.time.Event;
 import wres.datamodel.time.TimeSeries;
 import wres.datamodel.time.TimeSeriesMetadata;
 import wres.reading.DataSource;
+import wres.reading.ReadException;
 import wres.reading.TimeSeriesTuple;
 import wres.reading.DataSource.DataDisposition;
 import wres.statistics.MessageUtilities;
@@ -631,6 +634,66 @@ class PublishedInterfaceXmlReaderTest
                 List<TimeSeries<Double>> expected = List.of( expectedOne );
 
                 assertEquals( expected, actual );
+            }
+
+            // Clean up
+            if ( Files.exists( xmlPath ) )
+            {
+                Files.delete( xmlPath );
+            }
+        }
+    }
+
+    // GitHub 494
+    @Test
+    void testReadObservationsWithTimeZoneThatConflictsWithDeclarationProducesExpectedException() throws IOException
+    {
+        try ( FileSystem fileSystem = Jimfs.newFileSystem( Configuration.unix() ) )
+        {
+            // Write a new pi-xml file to an in-memory file system
+            Path directory = fileSystem.getPath( TEST );
+            Files.createDirectory( directory );
+            Path pathToStore = fileSystem.getPath( TEST, TEST_XML );
+            Path xmlPath = Files.createFile( pathToStore );
+
+            try ( BufferedWriter writer = Files.newBufferedWriter( xmlPath ) )
+            {
+                writer.append( PI_STRING_ONE );
+            }
+
+            DataSource dataSource = Mockito.mock( DataSource.class );
+            Mockito.when( dataSource.getUri() )
+                   .thenReturn( xmlPath.toUri() );
+            Mockito.when( dataSource.getVariable() )
+                   .thenReturn( new Variable( QINE, null, Set.of() ) );
+            Mockito.when( dataSource.hasSourcePath() )
+                   .thenReturn( true );
+            Mockito.when( dataSource.getDisposition() )
+                   .thenReturn( DataDisposition.XML_PI_TIMESERIES );
+
+            Dataset dataset = Mockito.mock( Dataset.class );
+            Mockito.when( dataset.unit() )
+                   .thenReturn( "CFS" );
+            Mockito.when( dataSource.getContext() )
+                   .thenReturn( dataset );
+
+            Source source = Mockito.mock( Source.class );
+            Mockito.when( source.timeZoneOffset() )
+                   .thenReturn( ZoneOffset.ofHours( -1 ) );
+            Mockito.when( dataSource.getSource() )
+                   .thenReturn( source );
+
+            PublishedInterfaceXmlReader reader = PublishedInterfaceXmlReader.of();
+
+            // No reading yet, we are just opening a pipe to the file here
+            try ( Stream<TimeSeriesTuple> tupleStream = reader.read( dataSource ) )
+            {
+                // Now we trigger reading because there is a terminal stream operation. Each pull on a time-series
+                // creates as many reads from the file system as necessary to read that time-series into memory
+                ReadException actual = assertThrows( ReadException.class,
+                                                     tupleStream::toList );
+                assertTrue( actual.getMessage()
+                                  .contains( "does not match the 'timeZone'" ) );
             }
 
             // Clean up

